solr/core/src/java/org/apache/solr/update/VersionInfo.java

```diff
@@ -205,9 +215,59 @@ public VersionBucket bucket(int hash) {
     // Make sure high bits are moved down, since only the low bits will matter.
     // int h = hash + (hash >>> 8) + (hash >>> 16) + (hash >>> 24);
     // Assume good hash codes for now.
+    int slot = hash & (numBuckets - 1);
+    if (bucketMap != null) {
+      return getBucketFromMap(slot);
+    }
+    return getBucketFromArray(slot);
+  }
 
-    int slot = hash & (buckets.length - 1);
-    return buckets[slot];
+  private VersionBucket getBucketFromMap(int slot) {
+    VersionBucket bucket = bucketMap.get(slot);
+    if (bucket == null) {
+      synchronized (bucketsSync) {
+        int index = bucketMap.indexOf(slot);
```

Comment:
is this a potential NPE? if a thread is waiting on the lock while in the background the map turns into an array. thread gets the lock, does a `bucketMap.indexOf` but the bucketMap is null by the transformation.

Reply:
Yes, good catch.